File: FundComponents/ClimateDynamicsComponent.cs
// FUND - Climate Framework for Uncertainty, Negotiation and Distribution
// Copyright (C) 2012 David Anthoff and Richard S.J. Tol
// http://www.fund-model.org
// Licensed under the MIT license
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Esmf;

namespace Fund.Components.ClimateDynamics
{

    /// <summary>State for the climate component</summary>
    public interface IClimateDynamicsState
    {
        /// <summary>Total radiative forcing</summary>
        IParameter1Dimensional<Timestep, double> radforc { get; }

        /// <summary>Average global temperature</summary>
        IVariable1Dimensional<Timestep, double> temp { get; }

        /// <summary>LifeTempConst</summary>
        double LifeTempConst { get; }

        /// <summary>LifeTempLin</summary>
        double LifeTempLin { get; }

        /// <summary>LifeTempQd</summary>
        double LifeTempQd { get; }

        /// <summary>Climate sensitivity</summary>
        double ClimateSensitivity { get; }
    }

    public class ClimateDynamicsComponent
    {

        public void Run(Clock clock, IClimateDynamicsState state, IDimensions dimensions)
        {
            var s = state;
            var t = clock.Current;

            if (clock.IsFirstTimestep)
            {
                s.temp[t] = 0.20;
            }
            else
            {
                var LifeTemp = Math.Max(s.LifeTempConst + s.LifeTempLin * s.ClimateSensitivity + s.LifeTempQd * Math.Pow(s.ClimateSensitivity, 2.0), 1.0);

                var delaytemp = 1.0 / LifeTemp;

                var temps = s.ClimateSensitivity / 5.35 / Math.Log(2.0);

                // Calculate temperature          
                var dtemp = delaytemp * temps * s.radforc[t] - delaytemp * s.temp[t - 1];

                s.temp[t] = s.temp[t - 1] + dtemp;
            }
        }
    }
}

File: src/components/ClimateDynamicsComponent.jl
@defcomp climatedynamics begin
    # Total radiative forcing
    radforc = Parameter(index=[time])

    # Average global temperature
    temp = Variable(index=[time])

    # lifetempconst
    lifetempconst = Parameter()

    # lifetemplin
    lifetemplin = Parameter()

    # lifetempqd
    lifetempqd = Parameter()

    # Climate sensitivity
    climatesensitivity = Parameter(default=2.999999803762826)

    function run_timestep(p, v, d, t)
        
        if is_first(t)
            v.temp[t] = 0.20
        else
            LifeTemp = max(p.lifetempconst + p.lifetemplin * p.climatesensitivity + p.lifetempqd * p.climatesensitivity^2.0, 1.0)

            delaytemp = 1.0 / LifeTemp

            temps = p.climatesensitivity / 5.35 / log(2.0)

            # Calculate temperature
            dtemp = delaytemp * temps * p.radforc[t] - delaytemp * v.temp[t - 1]

            v.temp[t] = v.temp[t - 1] + dtemp
        end
    end
end
